src/main/ruby/truffleruby/core/truffle/time_operations.rb

```diff
@@ -91,5 +91,29 @@ def self.compose(time_class, utc_offset, p1, p2 = nil, p3 = nil, p4 = nil, p5 =
 
       Primitive.time_s_from_array(time_class, sec, min, hour, mday, month, year, nsec, is_dst, is_utc, utc_offset)
     end
+
+    def self.new_from_string(time_class, str, **options)
+      raise ArgumentError, 'time string should have ASCII compatible encoding' unless str.encoding.ascii_compatible?
+
+      # Fast path for well-formed strings.
+      # Specify the acceptable range for each component in the regexp so that if
+      # any value is out of range the match will fail and fall through.
+      if match = str.match(/\A(\d{4,5})(?:-(0[0-9]|1[012])-(0[0-9]|[12][0-9]|3[01])[ T]([01][0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]|60)(?:\.(\d+))?\s*(\S+)?)?\z/)
```

Comment:
I think this regexp would be more readable if it included capture names like:
```ruby
if /\A(?<year>\d{4,5})(?:-(?<month>0[0-9]|1[012]))/ =~ str
  # year, month, etc are already local variables here
```

Using `//x` might also be worthwhile to use multiple lines, but might be unnecessary when the Regexp is simplified to not check the precise range of arguments.

Reply:
Agreed, looks nicer.